File: Java/DFS/Tree/PathSum.java
package DFS.Tree;

import DataStructures.TreeNode;

public class PathSum {
  // Leetcode 112: https://leetcode.com/problems/path-sum/description/
  // Submission Detail: https://leetcode.com/problems/path-sum/submissions/1648623536

  // Tx = O(n)
  // Sx = O(n)
  public boolean hasPathSum(TreeNode node, int targetSum) {
    // Base Case: Empty Node doesn't have target sum.
    if(node == null)
      return false;

    // Base Case: Leaf Node
    if(node.left == null && node.right == null)
      return node.val == targetSum;

    boolean left = hasPathSum(node.left, targetSum-node.val);
    boolean right = hasPathSum(node.right, targetSum-node.val);

    return left || right;
  }
}

File: C#/Leetcode/Patterns/DFS/PathSum.cs
﻿using LeetcodeSolutions.DataStructures;

namespace LeetcodeSolutions.Patterns.DFS
{
    // Leetcode 112 - https://leetcode.com/problems/path-sum/description/
    // Submission Detail - https://leetcode.com/submissions/detail/141729452/

    public class PathSum
    {
        // Tx = O(n)
        // Sx = O(d)    // for call stack
        public bool HasPathSum(BinaryTreeNode root, int sum)
        {
            if (root == null) return false;
            if (root.Left == null && root.Right == null) return root.Val == sum;

            return HasPathSum(root.Left, sum - root.Val) || HasPathSum(root.Right, sum - root.Val);
        }
    }
}
